modules/standard/Version.chpl

```diff
@@ -0,0 +1,349 @@
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+
+This module contains features that support compile-time reasoning
+about version numbers in general, and Chapel version numbers
+specifically.  In more detail, it features:
+
+* :var:`chplVersion`: the version number of the copy of ``chpl`` used
+  to compile the program.
+
+* two types that can be used to represent version numbers:
+
+  * :type:`semanticVersion` which represents a semantic version value
+
+  * :type:`sourceVersion` which is a semantic version plus a commit value
+
+* :proc:`version`: a utility function for creating new version values
+
+Version types in this module are defined in terms of `param` values to
+support compile-time reasoning about versions and code specialization.
+
+The version types support:
+
+* being printed out or cast to a ``param`` string
+
+* compile-time comparisons between version values via ``==``, ``!=``,
+  ``<``, ``<=``, ``>``, and ``>``, including mixed types.  Generally
+  speaking, "less than" corresponds to "is an earlier version than."
+
+*/
+
+module Version {
+  // Query the compiler to get its version number and cache those values
+  private param chplMajor = __primitive("version major");
+  private param chplMinor = __primitive("version minor");
+  private param chplUpdate = __primitive("version update");
+  private param chplSHA = __primitive("version sha");
+
+
+  /*
+    The version of ``chpl`` used to compile this program.  For an
+    official Chapel release, this will have type ``semanticVersion``
+    while for a pre-release, it will have type ``sourceVersion``.
+
+    Note that Chapel ``1.x.y``/``2.x.y`` corresponds to version
+    ``0.x.y``/``1.x.y`` in traditional semantic versioning.
+  */
+
+  const chplVersion;
+  chplVersion = version(chplMajor, chplMinor, chplUpdate, chplSHA);
+
+
+  /*
+    A helper function that creates a new version value from its
+    arguments.
+
+    :arg major: The major version number
+    :type major: `int`
+
+    :arg minor: The minor version number
+    :type minor: `int`
+
+    :arg update: The optional update version number (defaults to 0)
+    :type update: `int`
+
+    :arg commit: The optional commit ID (defaults to "")
+    :type commit: `string`
+
+    :returns: A new version value.  If commit == "", this routine
+              returns a value of :type:`semanticVersion` otherwise it
+              returns a value of :type:`sourceVersion`.
+  */
+  proc version(param major: int, param minor: int,
```

Comment:
I'd still be in favor of giving `minor` a default value so down the road you can do things like: `if chplVersion > version(2)`

Reply:
Whoops, I had missed your original response in https://github.com/chapel-lang/chapel/pull/13191#discussion_r436162130. I can see that argument, but I was thinking that there's a lot of times when I'm comparing gcc versions where all I care about is the major version. That said, having to write `if chplVersion > version(2, 0)` isn't much more work, so maybe that's not worthwhile.